gym_xiangqi/xiangqi_game.py

```diff
@@ -0,0 +1,144 @@
+        "king": King,
+        "queen": Queen,
+        "bishop": Bishop,
+        "knight": Knight,
+        "rook": Rook,
+        "cannon": Cannon,
+        "pawn": Pawn
+    }
+    pieceCount = {              # mapping of piece name to piece counts
+        "king": 1,
+        "queen": 2,
+        "bishop": 2,
+        "knight": 2,
+        "rook": 2,
+        "cannon": 2,
+        "pawn": 5
+    }
+    agentCoord = {              # mapping of piece name to agent initial positions
+        "king": [(9, 4)],
+        "queen": [(9, 3), (9, 5)],
+        "bishop": [(9, 2), (9, 6)],
+        "knight": [(9, 1), (9, 7)],
+        "rook": [(9, 0), (9, 8)],
+        "cannon": [(7, 1), (7, 7)],
+        "pawn": [(6, 0), (6, 2), (6, 4), (6, 6), (6, 8)]
+    }
+    enemyCoord = {              # mapping of piece name to enemy initial positions
+        "king": [(0, 4)],
+        "queen": [(0, 3), (0, 5)],
+        "bishop": [(0, 2), (0, 6)],
+        "knight": [(0, 1), (0, 7)],
+        "rook": [(0, 0), (0, 8)],
+        "cannon": [(2, 1), (2, 7)],
+        "pawn": [(3, 0), (3, 2), (3, 4), (3, 6), (3, 8)]
+    }
+
+    def __init__(self):
+        # Xiangqi components
+        self.board = [[None for _ in range(self.boardCols)] for _ in range(self.boardRows)]
+        self.agentColor = random.randint(0, 1)
+        self.enemyColor = 0 if self.agentColor == 1 else 1
+        self.initAllPieces()
+
+        # PyGame components
+        self.running = False
+        self.winWidth = 900
+        self.winHeight = 1000
+        self.dim = (self.winWidth, self.winHeight)
+        self.display_surf = None
+
+
+    def initAllPieces(self):
+        # initialize agent and enemy pieces and place them on the board
+        for piece, pieceClass in self.classMap.items():
+            for i in range(self.pieceCount[piece]):
+                r = self.agentCoord[piece][i][0]
+                c = self.agentCoord[piece][i][1]
+                self.board[r][c] = pieceClass(self.agentColor, row=r, col=c)
+
+                r = self.enemyCoord[piece][i][0]
+                c = self.enemyCoord[piece][i][1]
+                self.board[r][c] = pieceClass(self.enemyColor, row=r, col=c)
+
+    def onInit(self):
+        """
+        Initialize/start the game with PyGame
+        ex. pygame.init()
+            (anything related to PyGame module that needs to be initialized)
+        """
+        pygame.init()
+        self.display_surf = pygame.display.set_mode(self.dim, pygame.HWSURFACE | pygame.DOUBLEBUF)
+        return True
+
+    def onEvent(self, event):
+        """
+        This routine is triggered when some kind of user/game event is detected
+        ex. when user closes the PyGame window
+            (mostly any keyboard/mouse input)
+        """
+        if event.type == pygame.QUIT:
+            self.running = False
+
+    def onUpdate(self):
+        """
+        Relfect detected changes and update game states
+        """
+        pass
+
+    def render(self):
```

Comment:
I believe some of these methods (like render, reset) is not needed, since it's already in xiangqi_env.py. Do you plan to call these class method within the environment methods?

Reply:
Just FYI, I mean here: https://github.com/tanliyon/gym-xiangqi/blob/main/gym_xiangqi/envs/xiangqi_env.py

Reply:
I want to leave them for now because I really want to modularize the two components PyGame and Gym environment. So that PyGame only takes care the game itself and the environment only takes care the environment itself. In case we choose to implement these functions within our environment script, it will be easy to remove them anyways. My goal here was to layout the overall outline.